### ISSUE TITLE
Fire attributeChanged or equivalent for attributes not known upfront

### ISSUE THREAD

**Issue by ergo** [Polymer/polymer#4636]
Hello,

I'm working on a router element, that also has url generator.

It was working using markup like this: 
`<uirouter-sref  param-foo="[[fooVal]]" param-bar="[[barBinding]]" param-baz="Z">`

The parameters are not known upfront.

In the past for Polymer 1.0 I was able to add `attributeChanged` callback that worked on any attribute, whenever anything changed I could update the link/DOM easly.

In Polymer 2.0/WebComponents `observedAttributes` must explictly list all attributes that can be observed and this behavior breaks my element. Since data bindings in polymer do not allow very complex structures I was wondering if there is anything I could use to make this work.
If this is not possible, I ask on advice how to implement element like that that could build accept equivalent of named attributes that would work well with databinding.

**Comment by jfrazzano** [comment]
I have used hyper and hyper html.  fast, straight forward, updating bind on change.  DONT LEAVE SPACES IN YOUR DOM.

https://github.com/WebReflection/hyperHTML/tree/master/test <https://github.com/WebReflection/hyperHTML/tree/master/test>

https://github.com/WebReflection/viperHTML <https://github.com/WebReflection/viperHTML>

Its a work around that steps a little outside the traditional custom element approach. Not sure its what you want. 
I definitely went through the same sorts of issues. It was the best solution I could find, and Andreas—the library’s author--has been super helpful and supportive. 
I know its not exactly what you were looking for, but it’s the best I could come up with as of now.

Have a good evening.

Sincerely, 
Jason

**Comment by kevinpschaaf** [comment]
As noted, in V1 custom elements, there is no way to opt-in to receiving all attribute changes, they must be included in the `observedAttributes` filter.

Note however, `attributeChangedCallback` is not required for setting up data-binding, since those are parsed from the template once by the host of the element, and then set to _properties_ on the target element where property accessors are used to react to changes, as opposed to `attributeChangedCallback`.  As such, if you constrain the element's API to only accept attributes as static configuration once, you could write you element to scan its `this.attributes` list at `connectedCallback` time to pull initial values out (note, attributes are not available in`constructor` per spec), and then you can create accessors/effects for the properties to handle changes going forward.

I made a quick demo where the router only uses the `PropertyAccessors` mixin rather than the full `Polymer.Element`, since it may be simpler to just use that subset.  It's annotated with comments, feel free to ping me or re-open if you have questions.

**Comment by ergo** [comment]
I needed to react on binding changes too at later point, so i went with mutation observer to work around the problem for now. Would approach suggested by you be more performant by chance?

**Comment by ergo** [comment]
I'm currently playing with this and I think generally router should use properties by default. I'm trying to get this to work with a hybrid element (my current element also supports polymer 1.x so it would be great to keep it that way for now). Do I need to do something special to make it work in Polymer 1?

**Comment by ergo** [comment]
I got everything to work with a mutation observer. Not sure about performance implication for that though.

**Comment by ergo** [comment]
@kevinpschaaf Seems my conclusion about things working is a bit premature, there is one case that I'm not sure how to solve yet.

http://plnkr.co/edit/frEX6gp86h85B2G0Cs5a - here is a plunker (glitch seems broken atm) - the problem is that if some property does not have any value at the beginning (in this case `testBinding`), then the `_propertiesChanged` will not fire at all. When testBinding is uncommented in constructor things work at expected.

**Comment by kevinpschaaf** [comment]
Yeah, unfortunately I think it's just a chicken-egg situation.  You need to create an accessor to be able to see property changes, but you need the property set once to know the name of the accessor to create.  So you have to pick a point in time to look at the own properties that have been set to create the accessors, and anything set after that won't have accessors.

Bottom line, the router needs to know the properties it will create accessors for in order to make changes to those properties observable.  You could require the user to give all the property names at connected time, e.g. `params="foo bar test"` and create accessors for them (albeit this is unwieldy API-wise).  Alternatively you could require the user to give you the set of all params as a structured object, so there is only one well-known accessor to create (e.g. `params`).

**Comment by ergo** [comment]
Yeah, thats exactly what I did - apart from sniffing `param-*` I support `params`, but since polymers default template system allows for only very basic values ie. I can't pass `{foo, bar, baz}` without resorting to computed binding, I tried using this approach as a band aid, using `foo$="[[bar]]"` works okish for simple values. When `lit-html` becomes stable this will get easier to use and the problem will be remedied I hope.
